File: 0013-roman-to-integer/0013-roman-to-integer.py
class Solution(object):
    def romanToInt(self, s):
        roman_values = {
            'I': 1, 'V': 5, 'X': 10, 'L': 50, 
            'C': 100, 'D': 500, 'M': 1000
        }
        
        total = 0
        dk = 0  

        for char in reversed(s):
            curr_value = roman_values[char]
            if curr_value < dk:
                total -= curr_value
            else:
                total += curr_value
            dk = curr_value

        return total

File: 0013-roman-to-integer/0013-roman-to-integer.java
class Solution {
    public int romanToInt(String s) {
        
        HashMap<Character, Integer> hm = new HashMap<>();
        // RomanMap
        // refer Note

        hm.put('I', 1);
        hm.put('V', 5);
        hm.put('X', 10);
        hm.put('L', 50);
        hm.put('C', 100);
        hm.put('D', 500);
        hm.put('M', 1000);
        
        int prev = 0;
        int total = 0;

        for(int i = s.length() - 1; i >= 0; i--){
            int curr = hm.get(s.charAt(i));
            
            if(curr < prev)
                total -= curr;

            else
                total += curr;

            prev = curr;  

        }

        return total;


    }
}